File: Codes/Bingran Li/Tree/257. Binary Tree Paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        
        String myString = new String();
        List myList = new ArrayList();
        subPath(root,myString,myList);
        return myList;
        
    }
    public void subPath(TreeNode root, String tem,List myList){
       
        // solution 1: recursive
        if(root==null)   return;
        // method 1: in this way it create a new string, not affect old one
        // s += "->" + Integer.toString(root.val);
        // method 2: create a new string
         String s = tem+"->" + Integer.toString(root.val); // a safe way
        // method 3: (based on the original one)
        // this "s.concat(myString)" will return a new string, you need to use a new string to hold it
        // from s.concat() to tem=s.concat();
        if(root.left!=null)    subPath(root.left,s,myList);
        if(root.right!=null)   subPath(root.right,s,myList);
        // then left and right both are null
        if(root.right==null && root.left==null)  {
           
             myList.add(s.substring(2));
        
        }
       
    }
    
    
    
}

File: Codes/Kan Zheng/C++/DFS/257. Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> res;
        string s;
        dfs(res, s, root);
        return res;
    }
    
    void dfs(vector<string>& res, string& s, TreeNode* root) {
        if(!root) {
            return;
        }
        string str = to_string(root->val);
        if(!root->left && !root->right) {
            s += str;
            res.push_back(s);
            s.erase(s.size() - str.size());
        }else {
            s += str + "->";
            dfs(res, s, root->left);
            dfs(res, s, root->right);
            s.erase(s.size() - str.size() - 2);
        }
    }
};